app/public/cantusdata/helpers/mei_conversion/mei_parse.py

```diff
@@ -27,148 +19,108 @@
+    if semitones < 0:
         return "d"
-    elif prev_oct < cur_oct:
+    elif semitones > 0:
         return "u"
     else:
-        cur_pitch_value = PITCH_VALUES[cur_note.lower()]
-        prev_pitch_value = PITCH_VALUES[prev_note.lower()]
-        if prev_pitch_value > cur_pitch_value:
-            return "d"
-        elif prev_pitch_value < cur_pitch_value:
-            return "u"
-        else:
-            return "s"
-
-def get_streams(ncs: list) -> tuple:
-    contour = ""
-    pitches = ""
-    intervals = ""
-    prev = None
-    for nc in ncs:
-        #contour
-        if prev != None:
-            contour += contour_from_component_pair((prev.get("pname"), prev.get("oct")), (nc.get("pname"), nc.get("oct")))
-        #pitch
-        pitches += nc.get("pname").upper() + nc.get("oct") + ","
-        #interval
-            #translate pitches to numbers (whole or semitones) (and add 8*octave?)
-            #subtract, +/-
-        prev = nc
-    
-    return (contour, pitches, intervals)
+        return "s"
 
-def parse(file):
-    tree = etree.parse(file)
-    music = tree.getroot()[1]
 
+def parse_zones(mei):
+    """Get the zones (bounding boxes) from an MEI root element."""
     zones = {}
-    surface = music[0][0]
-    for zone in surface:
-        zone_id = zone.get('{http://www.w3.org/XML/1998/namespace}id')
-        ulx = int(zone.get("ulx", 0))
-        uly = int(zone.get("uly", 0))
-        lrx = int(zone.get("lrx", 0))
-        lry = int(zone.get("lry", 0))
-        rotation = zone.get("rotation", "0")
-        region = f"{ulx},{uly},{lrx-ulx},{lry-uly}"
-        zones["#"+zone_id] = {"region": region, "rotation": rotation}
-
-    layer = music[1][0][0][1][0][0]
+    for zone in mei.iter(f"{MEINS}zone"):
+        zone_id = zone.get(f"{XMLNS}id")
+        coordinate_names = ["ulx", "uly", "lrx", "lry"]
+        coordinates = [int(zone.get(c, 0)) for c in coordinate_names]
+        rotate = float(zone.get("rotate", 0.0))
+        zones[f"#{zone_id}"] = {
+            "coordinates": tuple(coordinates),
+            "rotate": rotate,
+        }
+    return zones
+
+
+def parse_neumes(mei, zones):
+    """Get all neume groupings and global pitch|interval|contour sequences."""
+    neumes = []
+    global_pitches = []
+    for syllable in mei.iter(f"{MEINS}syllable"):
+        syl = syllable.find(f"{MEINS}syl")
+        syltext = syl.text.strip()
+        syl_facs = syl.get("facs")
+        syl_coordinates = zones.get(syl_facs, (-1, -1, -1, -1))
+        for neume in syllable.findall(f"{MEINS}neume"):
+            pitches = []
+            nc_coordinates = []
+            for nc in neume:
+                pname = nc.get("pname")
+                oct = nc.get("oct")
+                facs = nc.get("facs")
+                coordinates = zones.get(facs, (-1, -1, -1, -1))
+                pitches.append(f"{pname}{oct}")
+                nc_coordinates.append(coordinates)
+            global_pitches.extend(pitches)
```

Comment:
so when you match a substring, you get something like `substring_idx_start` and `substring_idx_end`... then your search the neume grouping

```python
initial_neume_idx = global_sequences["neume_groupings"][substring_idx_start]
last_neume_idx = global_sequences["neume_groupings"][substring_idx_end]
...

initial_neume = neumes[initial_neume_idx]
last_neume = neumes[last_neume_idx]
```

not quite, but hopefully you get the idea

Reply:
No, no, you're right. It's just as simple as adding 1 to the last index so we get the _second_ pitch of the last interval. In case the interval spans two neumes we want to return both, right